dockerfiles/theia/Dockerfile

```diff
@@ -10,33 +10,44 @@
 FROM node:8-alpine
 
 ARG GITHUB_TOKEN
-ARG THEIA_VERSION=0.3.10
-ENV USE_LOCAL_GIT=true \
-    GITHUB_TOKEN=${GITHUB_TOKEN} \
+ARG THEIA_VERSION=0.4.0-next.b17727c1
+
+EXPOSE 3000 3030
+
+ENV GITHUB_TOKEN=${GITHUB_TOKEN} \
     THEIA_VERSION=${THEIA_VERSION} \
-    YARN_CACHE_FOLDER=/usr/local/share/.cache/yarn
+    USE_LOCAL_GIT=true \
+    HOME=/home/theia
 
-# build dependencies requireed to compile a custom Theia
-RUN apk add --no-cache make gcc g++ python git openssh bash supervisor jq
-WORKDIR /home/theia
-# build Theia with all extensions to persist yarn cache in the image and
-# have default Theia build in the workspace in case no plugins are requested
-ADD https://raw.githubusercontent.com/theia-ide/theia/v${THEIA_VERSION}/examples/browser/package.json /home/theia/package.json
-ADD theia-default-package.json /home/default/theia/package.json
-ADD versions.sh /home/default/versions.sh
+ADD src/ ${HOME}/
 ADD supervisord.conf /etc/
-RUN cd /home/default && ./versions.sh && \
-    cd /home/theia && \
-    yarn && \
-    yarn theia build --mode development && \
-    rm -rf * && \
-    cd /home/default/theia && \
+
+RUN \
+    # Install needed software
+    apk update && apk add --no-cache make gcc g++ python git bash supervisor jq && \
+    rm -rf /tmp/* /var/cache/apk/* && \
+    # Change version of Theia to specified in THEIA_VERSION
+    ${HOME}/versions.sh && rm ${HOME}/versions.sh && \
+    # Add Theia extensions
+    #==TODO== replace with Eclipse official version after Theia release
+    #git clone https://github.com/eclipse/che-theia-hosted-plugin-manager-extension /tmp/hosted-plugin-extension && \
+    git clone https://github.com/mmorhun/che-theia-hosted-plugin-manager-extension /tmp/hosted-plugin-extension && \
+    node ${HOME}/add-extensions.js \
+        @eclipse-che/che-theia-hosted-plugin-manager-extension:file:///tmp/hosted-plugin-extension \
+        #che-theia-ssh-extension:https://github.com/eclipse/che-theia-ssh-plugin.git \
+    && rm ${HOME}/add-extensions.js && \
+    # Build Theia with all the extensions
+    cd ${HOME} && \
+    cat package.json && \
     yarn && \
-    yarn theia build
-RUN find ${YARN_CACHE_FOLDER} -exec sh -c "chgrp 0 {}; chmod g+rwX {}" \;
-RUN mkdir -p /.yarn && chgrp 0 /.yarn && chmod g+rwX /.yarn && \
-    touch /.yarnrc && chgrp 0 /.yarnrc && chmod g+rwX /.yarnrc
-ADD src/main.js /theia_launcher/theia_launcher.js
-EXPOSE 3000
+    yarn theia build && \
+    # Install Theia plugin generator
+    #==TODO== add after generator-theia-plugin release
+    #npm install -g yo @wiptheia/generator-theia-plugin && \
```

Comment:
you think we shouldn't include the wip package for now ?

Reply:
I think we need release it first. Do not want to download wiptheia core and stuff.

Reply:
ah yes I forgot to push changes that use @theia/plugin
let's keep commented as you made 👍